crates/nu-parser/src/flatten.rs

```diff
@@ -668,20 +560,52 @@ pub fn flatten_pattern(match_pattern: &MatchPattern) -> Vec<(Span, FlatShape)> {
                 output.push((match_pattern.span, FlatShape::MatchPattern));
             }
         }
-        Pattern::Value(_) => {
-            output.push((match_pattern.span, FlatShape::MatchPattern));
-        }
-        Pattern::Variable(var_id) => {
-            output.push((match_pattern.span, FlatShape::VarDecl(*var_id)));
-        }
-        Pattern::Rest(var_id) => {
-            output.push((match_pattern.span, FlatShape::VarDecl(*var_id)));
-        }
+        Pattern::Value(_) => output.push((match_pattern.span, FlatShape::MatchPattern)),
+        Pattern::Variable(var_id) => output.push((match_pattern.span, FlatShape::VarDecl(*var_id))),
+        Pattern::Rest(var_id) => output.push((match_pattern.span, FlatShape::VarDecl(*var_id))),
         Pattern::Or(patterns) => {
             for pattern in patterns {
-                output.extend(flatten_pattern(pattern));
+                flatten_pattern_inner(pattern, output);
             }
         }
     }
+}
+
+pub fn flatten_block(working_set: &StateWorkingSet, block: &Block) -> Vec<(Span, FlatShape)> {
+    let mut output = Vec::new();
+    flatten_block_inner(working_set, block, &mut output);
+    output
+}
+
+pub fn flatten_pipeline(
+    working_set: &StateWorkingSet,
+    pipeline: &Pipeline,
+) -> Vec<(Span, FlatShape)> {
+    let mut output = Vec::new();
+    flatten_pipeline_inner(working_set, pipeline, &mut output);
+    output
+}
+
+pub fn flatten_pipeline_element(
+    working_set: &StateWorkingSet,
+    pipeline_element: &PipelineElement,
+) -> Vec<(Span, FlatShape)> {
+    let mut output = Vec::new();
+    flatten_pipeline_element_inner(working_set, pipeline_element, &mut output);
+    output
+}
+
+pub fn flatten_expression(
+    working_set: &StateWorkingSet,
+    expr: &Expression,
+) -> Vec<(Span, FlatShape)> {
+    let mut output = Vec::new();
+    flatten_expression_inner(working_set, expr, &mut output);
+    output
+}
+
+pub fn flatten_pattern(match_pattern: &MatchPattern) -> Vec<(Span, FlatShape)> {
+    let mut output = Vec::new();
+    flatten_pattern_inner(match_pattern, &mut output);
```

Comment:
In my brief goto reference checking `flatten_pattern`, `flatten_pipeline` and `flatten_expression` are not used outside this file, so they mabye don't need a public version with the `Vec::new`?

Reply:
Yeah I thought about that. Since `flatten_pipeline_element` is used externally, I'm inclined to keep the other ones for consistency/completeness in case something somewhere needs these functions. But I guess we can also just remove them.

Reply:
External as in a plugin? Or just in the completer? Kind of curious why that is the level of segmentation chosen there.

Reply:
External, as in external to the `nu-parser` crate. I.e., the fact that we consume `flatten_pipeline_element` from another one of our crates indicates to me that other crates (3rd party or one of ours) might want to use to the other flattening functions.

Reply:
Generally I am always inclined to minimize the public interface to reduce our stability risk and find dead code to eliminate. But it seems we maybe want to keep more than those required by the workspace.

Reply:
Agreed, maybe I should get rid `flatten_pattern`, since I think I added that one in this PR.

Reply:
In another PR, I can try to remove our usage of `flatten_pipeline_expression` and then remove it, `flatten_pipeline`, and `flatten_expression` from the public API.

Reply:
I am not saying you need to ditch the `flatten_pipeline_element` in nu-cli, but I question that we should expose anything for which we don't have a current use, and only make it pub on articulated need.

Reply:
I see, that depends on how we are treating `nu-parser`. Since it is listed on crates.io, I was treating it like a library. I.e, either the function is useful enough to be exported and to justify the maintenance burden, or it is not. I don't think whether we use it or not is the only factor that determines whether or not it should be public, since others may have valid use cases. E.g., if we were to stop using `flatten_pipeline_element` and start using `flatten_pipeline` instead, does that mean one should be removed from the public API in favor of the other?

Reply:
I think the larger library policy of our "internal" crates is outside the scope of this PR, so I won't push this further here.

Any missing steps to turn this draft into a merge?

Reply:
I might look it over one more time, but otherwise it should be ready.